File: backend/src/main/java/dev/frilly/locket/data/Friendship.java
package dev.frilly.locket.data;

import dev.frilly.locket.data.key.FriendshipKey;
import jakarta.persistence.*;

/**
 * Entity model class for a friendship between two users.
 */
@Entity
@Table(
    name = "friendships",
    uniqueConstraints = {@UniqueConstraint(columnNames = {"user1", "user2"})}
)
public final class Friendship {

  @EmbeddedId
  private final FriendshipKey key;
  @ManyToOne
  @MapsId("user1")
  @JoinColumn(name = "user1", nullable = false)
  private User user1;
  @ManyToOne
  @MapsId("user2")
  @JoinColumn(name = "user2", nullable = false)
  private User user2;

  public Friendship() {
    key = new FriendshipKey();
  }

  public Friendship(User user1, User user2) {
    key        = new FriendshipKey(user1.id(), user2.id());
    this.user1 = user1;
    this.user2 = user2;
  }

  public User user1() {
    return user1;
  }

  public User user2() {
    return user2;
  }

}

File: backend/src/main/kotlin/dev/frilly/locket/data/Friendship.kt
package dev.frilly.locket.data

import jakarta.persistence.*
import java.io.Serializable

/**
 * Composite primary key for the friendship table.
 */
@Embeddable
data class FriendshipCompositeKey(
    var user1: Long,
    var user2: Long,
) : Serializable {

    init {
        val temp = user1
        if (user1 > user2) {
            user1 = user2
            user2 = temp
        }
    }

}

/**
 * A relationship between two users.
 *
 * A table has:
 * - USER1 (Serial) primary key
 * - USER2 (Serial) primary key
 */
@Entity
@Table(
    name = "friendships",
    uniqueConstraints = [UniqueConstraint(columnNames = ["user1", "user2"])],
)
data class Friendship(
    @EmbeddedId
    val id: FriendshipCompositeKey,

    @ManyToOne
    @MapsId("user1")
    @JoinColumn(name = "user1", nullable = false)
    val user1: User,

    @ManyToOne
    @MapsId("user2")
    @JoinColumn(name = "user2", nullable = false)
    val user2: User,
)
